### ISSUE TITLE
[Feature Request] Easier way to use Pyro models as surrogates

### ISSUE THREAD

**Issue by exs-hkenlay** [pytorch/botorch#1714]
# 🚀 Feature Request

I want to use a Pyro model as a surrogate to do Bayesian optimisation in BoTorch. 

In the [docs page for Models ](https://botorch.org/docs/models#implementing-custom-models) I see that

>  The only requirement for using BoTorch's Monte-Carlo based acquisition functions is that the model has a posterior method. It takes in a Tensor X of design points, and returns a Posterior object describing the (joint) probability distribution of the model output(s) over the design points in X

It also gives an example 

> See the [Using a custom BoTorch model in Ax](https://botorch.org/tutorials/custom_botorch_model_in_ax) tutorial for an example.

Q1: the example does not implement a `posterior` method, but rather a `forward` method. Is that a typo in docs? 

The example uses GPyTorch and the Posterior object is from the `gpytorch.distributions.MultivariateNormal` which implements `rsample`. However, pyro distributions (thin wrappers around torch distributions) do not implement this so we cannot take the same approach.

Q2 (feature request): I was wondering if a class `PyroPosterior` in the same spirit of `botorch.posteriors.torch.TorchPosterior` would be a sensible way to implement custom Pyro models. The class would wrap a pyro posterior object but abide by the `botorch.posteriors.posterior.Posterior` API, then allowing easy design of the forward method described in the docs? 

Q3:As an aside, I noticed botorch.posteriors.torch.TorchPosterior` does not implement `base_sample_shape`, is this a bug?

Q4: I noticed there is a `PyroModel` class in `botorch.models.fully_bayesian` but it looks like its there solely to help build the `SaasFullyBayesianSingleTaskGP` class which also returns a `gpytorch.distributions.MultivariateNormal`. Have I misunderstood the `PyroModel` class?

Q5: I assume this it would not be hard to implement this to also allow back-propagating gradients through the samples to the model input?

## Motivation

It should be easy to use Pyro models for Monte-Carlo based acquisition functions with derivatives.

## Pitch

**Describe the solution you'd like**
A `PyroPosterior` class that takes a Pyro posterior and gives a posterior useable in BoTorch.

**Describe alternatives you've considered**
I tried using Pyro distributions but they do not satisfy the BoTorch API. The name of the `PyroModel` class suggests an option to take a Pyro model and turn it into a BoTorch model but as far as I can tell this is not the case.

**Are you willing to open a pull request?** (See [CONTRIBUTING](../CONTRIBUTING.md))
Yes, although it may take me some weeks

**Comment by hkenlay** [comment]
```
import botorch
import pyro
import torch
from typing import Optional
from torch import Tensor
import warnings

class PyroDistribution(botorch.posteriors.Posterior):
    """A wrapper around pyro distribution that fits the botorch distribution API."""

    def __init__(self, distribution: pyro.distributions.distribution.Distribution):
        super().__init__()
        self.distribution = distribution
        for attr in vars(self.distribution).values():
            if isinstance(attr, torch.Tensor):
                self._device = attr.device
                self._dtype = attr.dtype
                break

    @property
    def device(self) -> torch.device:
        return self._device

    @property
    def dtype(self) -> torch.dtype:
        return self._dtype

    def rsample_from_base_samples(
        self,
        sample_shape: torch.Size,
        base_samples: Tensor,
    ) -> Tensor:
        if base_samples.shape[: len(sample_shape)] != sample_shape:
            raise RuntimeError("`sample_shape` disagrees with shape of `base_samples`.")
        samples = self.distribution.rsample(
            sample_shape=sample_shape
        )
        samples = samples.unsqueeze(-1)
        return samples

    def rsample(
        self,
        sample_shape: Optional[torch.Size] = None,
        base_samples: Optional[Tensor] = None,
    ) -> Tensor:
        if sample_shape is None:
            sample_shape = torch.Size([1])
        if base_samples is not None:
            warnings.warn(
                "Use of `base_samples` with `rsample` is deprecated. Use "
                "`rsample_from_base_samples` instead.",
                DeprecationWarning,
            )
            if base_samples.shape[: len(sample_shape)] != sample_shape:
                raise RuntimeError(
                    "`sample_shape` disagrees with shape of `base_samples`."
                )
            # get base_samples to the correct shape
            base_samples = base_samples.expand(self._extended_shape(sample_shape))
            base_samples = base_samples.squeeze(-1)
            return self.rsample_from_base_samples(
                sample_shape=sample_shape, base_samples=base_samples
            )
        samples = self.distribution.rsample(
            sample_shape=sample_shape
        )
        samples = samples.unsqueeze(-1)
        return samples

    @property
    def base_sample_shape(self) -> torch.Size:
        return self.distribution.event_shape

    @property
    def batch_range(self) -> tuple[int, int]:
        return (0, -1)
```

Here is an example based on `botorch.posteriors.gpytorch.GPyTorchPosterior`. Would be happy to put in a PR with this.

**Comment by Balandat** [comment]
Hey, sorry for the long delay in the response. 

> Q1: the example does not implement a posterior method, but rather a forward method. Is that a typo in docs?

No it is not - the `GPyTorchModel` class that the custom model subclasses implements this `posterior` method, so we're inheriting from some generic work done for many GPyTorch models: https://github.com/pytorch/botorch/blob/main/botorch/models/gpytorch.py#L152

> Q2 (feature request): I was wondering if a class PyroPosterior in the same spirit of botorch.posteriors.torch.TorchPosterior would be a sensible way to implement custom Pyro models. The class would wrap a pyro posterior object but abide by the botorch.posteriors.posterior.Posterior API, then allowing easy design of the forward method described in the docs?

Yeah I think that would be great to have.

> Q3: As an aside, I noticed `botorch.posteriors.torch.TorchPosterior` does not implement `base_sample_shape`, is this a bug?

No, it is not. `base_sample_shape` is assuming samples from the posterior can be expressed as transformations of simpler base samples (the "reparameterization trick"). `TorchPosterior` doesn't require that this be the case, so it doesn't implement this functionality. 

> Q4: I noticed there is a PyroModel class in botorch.models.fully_bayesian but it looks like its there solely to help build the SaasFullyBayesianSingleTaskGP class which also returns a gpytorch.distributions.MultivariateNormal. Have I misunderstood the PyroModel class?

No, you did not misunderstand, this is correct. The idea of this is to have a pyro model run the MCMC chain to estimate the posterior hyperparameter distribution of a fully Bayesian model. We then draw a fixed number of samples from that distribution to load into a batched GPyTorch model. The actual model used for acquisition function optimization then doesn't use pyro at all anymore.  

> Q5: I assume this it would not be hard to implement this to also allow back-propagating gradients through the samples to the model input?

Yeah so I am not an expert on pyro, but so long as the samples drawn from the model posterior allow taking gradients w.r.t. the inputs to the model (i.e. if everything is implemented in a differentiable fashion) then we should be in business.


> Here is an example based on botorch.posteriors.gpytorch.GPyTorchPosterior. Would be happy to put in a PR with this

Hooray! Have you tried this on an actual pyro model? It would be great to see an working prototype end-to-end before adding this in (that could also serve as a nice tutorial / documentation for this new functionality).

**Comment by maxnpopov** [comment]
Hi! Is there any follow-up activity based on this request?

**Comment by Balandat** [comment]
Not on our end; @exs-hkenlay are you still planning to put a PR up for this?

**Comment by exs-hkenlay** [comment]
Unfortunately, its not a priority for me at the moment. I would love to come back to it in the future, but happy for the issue to be closed for now
